### ISSUE TITLE
[BUG]: AzureFileCopy task fails when using Workload Identity Federation

### ISSUE THREAD

**Issue by Sharparam** [microsoft/azure-pipelines-tasks#19207]
### Task name

AzureFileCopy

### Task version

5

### Environment type (Please select at least one enviroment where you face this issue)

- [ ] Self-Hosted
- [x] Microsoft Hosted
- [ ] VMSS Pool
- [ ] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

windows-2022

### Task log

```shell
Too long to be pasted: https://gist.github.com/Sharparam/c3bd654b31a9e6fa87473d89f71d2bdf
```

### Relevant log output

```shell
##[error]Upload to container: 'redacted' in storage account: 'redacted' with blob prefix: '' failed with error: 'Unsupported authentication scheme 'WorkloadIdentityFederation' for endpoint.' For more info please refer to https://aka.ms/azurefilecopyreadme
```

### Aditional info

```shell
v5 is the latest version of the task.
```

**Comment by anish714** [comment]
Running into the same issue.  Apparently it's the same issue for dacpacdeployment and it was fixed there: https://github.com/microsoft/azure-pipelines-tasks/issues/18992

**Comment by anish714** [comment]
Running into the same issue.

**Comment by anish714** [comment]
Also I guess this is not a bug.. lol https://developercommunity.visualstudio.com/t/not-enough-permissions-when-creating-aad-applicati/302555

**Comment by amitjaura** [comment]
Having same issue. Reverted back from Workload Identity Federation and it worked fine

**Comment by geekzter** [comment]
azcopy itself does not yet support Workload identity federation, see [issue](https://github.com/Azure/azure-storage-azcopy/issues/2112) in the azure-storage-azcopy repo. Once it does we will update the task as well.

**Comment by kaskeloten** [comment]
Having a similar issue with AzureFileCopy@3 using workload identity federation with openid connect, though It was working earlier today. 
Stopped working after the task got updated from 3.230.2 to 3.231.2, could be unrelated though.

**Comment by DekCode** [comment]
Has anyone got it working or any work around for this?

**Comment by Sharparam** [comment]
> Has anyone got it working or any work around for this?

We just switched back to the old authentication method for the time being.

**Comment by VassilisM** [comment]
Hi, same for me. Thing is... by searching around, I get what "WorkloadIdentityFederation" somewhat is, but I do not know how this is setup and therefore how to revert. Help, please?

**Comment by arsh3186** [comment]
Hi 
I am running into same issue, any updates on this? Is it working for anyone

**Comment by zee-cdp** [comment]
We are having same issue

**Comment by gsimardnet** [comment]
Workload Identity Federation is the recommended method to create a service connection. This task should be able to support it.

**Comment by stevenoneill** [comment]
Would love to see this updated.

**Comment by MattJeanes** [comment]
We've just been bitten by this as well. Workload identity federation is now the recommended way with prompts in the UI to upgrade so it was very much a surprise for us to see Microsoft maintained tasks still not supporting this on their newest task version.

**Comment by pilcherd** [comment]
Yep, us too. Set up a new "Recommended" connection and our build script didn't run and copy. Back to the old connection until there is a path forward.

**Comment by stevenoneill** [comment]
I followed [#2112](https://github.com/Azure/azure-storage-azcopy/issues/2112) through to https://github.com/Azure/azure-storage-azcopy/issues/2545.  2545 seems to have some good active updates -- keeping an eye on that.

**Comment by benjiecov** [comment]
The latest update on that azcopy issue says it's not currnetly a priority. But Azure DevOps now only creates/renews the secrets for service principal service connections for 3 months. So Azure DevOps is pushing us to use workload identity federation to avoid frequent secret management, while azcopy (and by extension this task) don't support it.

**Comment by dfederm** [comment]
Policies are now being enforced which require workload identities, so the `AzureFileCopy` pipeline task is now effectively entirely unuseable.

**Comment by dfederm** [comment]
I was able to work around this by ditching AzureFileCopy in favor of the Azure CLI task (which properly supports workload identities) with the "az storage blob upload-batch" command:

```
az storage blob upload-batch --account-name <blobAccountName> -d <containerName> -s <sourceDirectory> --overwrite true
```

**Comment by eriksteinebach** [comment]
@dfederm Same for us, it didn't allow me to update the service connection, so had to upgrade. Now this task is broken. I will give your fix a try, but the Create SAS Token and AzureFileCopy task now has become unusable for us

**Comment by vihanga-bandara** [comment]
We are getting the same issue and we use AzureFileCopy@3. As everyone else has mentioned this seems to be due to service connection supporting Workload Identity Federation but the task itself does not support it. Looking at the issue [](https://github.com/Azure/azure-storage-azcopy/issues/2545), they might not fix it for sometimes

**Comment by eriksteinebach** [comment]
I was able to solve it by using an AzureCLI task, maybe this will help some people:
```
steps:
- task: AzureCLI@2
  displayName: 'Copy ARM templates to artifact storage'
  inputs:
    azureSubscription: '<SUBSCRIPTION>'
    scriptType: pscore
    scriptLocation: inlineScript
    inlineScript: |
     az storage blob upload-batch --account-name "xartifacts" -d "artifacts" -s "$(System.DefaultWorkingDirectory)/_accountname/drop/Infrastructure/" --overwrite true
     
     $expiretime=$(date -u -d '30 minutes' +%Y-%m-%dT%H:%MZ)
     
     $connection=$(az storage account show-connection-string --resource-group "resourcegroupname" --name "xartifacts" --query connectionString)
     
     echo "##vso[task.setvariable variable=storageToken]?$(az storage container generate-sas --name artifacts --expiry $expiretime --permissions r --output tsv --connection-string $connection)"
     
     echo "##vso[task.setvariable variable=storageUri]https://xartifacts.blob.core.windows.net/artifacts"
```
This sets the two variables storageToken and storageUri to call the specific ARM template files I need

**Comment by mwiemer-microsoft** [comment]
v6 was released yesterday and resolves this issue for us :)

**Comment by geekzter** [comment]
The [AzureFileCopyV6](https://learn.microsoft.com/azure/devops/release-notes/2024/pipelines/sprint-236-update#new-azurefilecopy6-task-supports-secret-less-configurations) task resolves this issue. It also uses Azure RBAC to access blob storage instead. This requires the identity of the service connection used to have the appropriate RBAC role e.g. [Storage Blob Data Contributor](https://learn.microsoft.com/azure/role-based-access-control/built-in-roles/storage#storage-blob-data-contributor). See [Assign an Azure role for access to blob data](https://learn.microsoft.com/azure/storage/blobs/assign-azure-role-data-access?tabs=portal).

**Comment by Veera-kambham** [comment]
SASToken is not getting generated in this task. The options fields are missing

**Comment by aya-bjoseph** [comment]
FYI - For anyone who might have stumbled upon this issue, there appears to be an ongoing regression bug in AzureFileCopyV6 that is blocking use of WIF service connections. See issue #19566 and fix PR #19863 which hasn't been merged as yet.

cc: @geekzter

**Comment by v-schhabra** [comment]
> FYI - For anyone who might have stumbled upon this issue, there appears to be an ongoing regression bug in AzureFileCopyV6 that is blocking use of WIF service connections. See issue [#19566](https://github.com/microsoft/azure-pipelines-tasks/issues/19566) and fix PR [#19863](https://github.com/microsoft/azure-pipelines-tasks/pull/19863) which hasn't been merged as yet.
> 
> cc: [@geekzter](https://github.com/geekzter)

https://github.com/microsoft/azure-pipelines-tasks/pull/19867
This PR has been merged after incorporating the fixes.
